### PR TITLE
CKI: create a test wrapper for upstream kernel selftest

### PR DESCRIPTION
**Pull request by liuhangbin**
Currently we only support bpf, net, forwarding, tc tests

Signed-off-by: Hangbin Liu <haliu@redhat.com>

### CODE REVIEW

File: kselftests/include.sh [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+{
+	# clean environment
+	# restore environment
+	unset PASS
+	unset FAIL
+}
+
+log()
+{
+	#echo ":: [  LOG   ] :: : $1" | tee -a $OUTPUTFILE
```
Comment (emanchado) [inline kselftests/include.sh]: Can we remove commented-out lines?

File: kselftests/include.sh [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+# Usage: run command [return_value]
+run()
+{
+	cmd=$1
+	# FIXME: only support zero or none zero, doesn't support 2-10, or 2,3,4
+	exp=${2:-0}
+	echo -e "\n[$(date '+%T')][$(whoami)@$(uname -r | cut -f 2 -d-)]# '"$cmd"'" | tee -a $OUTPUTFILE
+	# FIXME: how should we handle if there are lots of output for the cmd,
+	# and we only care the return value
+	eval "$cmd" &> >(tee -a $OUTPUTFILE)
```
Comment (emanchado) [inline kselftests/include.sh]: Oh, I had never seen this. How does it work?

File: kselftests/include.sh [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+# Usage: watch command timeout [signal]
+watch()
+{
+	command=$1
+	timeout=$2
+	single=${3:-9}
+	now=`date '+%s'`
+	after=`date -d "$timeout seconds" '+%s'`
+
+        eval "$command" &
```
Comment (emanchado) [inline kselftests/include.sh]: Why the different indentation here, compared to the variables?

File: kselftests/include.sh [hunk 4]
```diff
@@ -0,0 +1,156 @@
+#!/bin/bash
+# include beaker default environmnet
+[ -f /usr/bin/rhts_environment.sh ] && source /usr/bin/rhts_environment.sh
```
[the remark this answers is not in the document]
Comment (liuhangbin) [reply kselftests/include.sh]: Beaker still uses it. But yes, I can change it to beakerlib. I need to check how to use beaker lib to report pass/fail first.

File: kselftests/include.sh (hunk 1 above)
Comment (liuhangbin) [reply kselftests/include.sh]: OK, I will

File: kselftests/include.sh (hunk 2 above)
Comment (liuhangbin) [reply kselftests/include.sh]: This is used for redirecting the cmd output to log file while we still could get the return code.

File: kselftests/include.sh (hunk 3 above)
Comment (liuhangbin) [reply kselftests/include.sh]: Sorry I didn't get your question. 

The run/watch functions are a re-write of rlRun and rlWatchdog. The purpose is to make user could run the test even without rhts/beakerlib.

File: kselftests/include.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (liuhangbin) [reply kselftests/include.sh]: Hi @emanchado , I checked the rhts- functions and it looks there is no relation with the old RHTS. The current rhts- functions are wrappers of restraint functions. So I think it should be OK to use it. On the other hand, as my tests do not have rlPhase, I could use rlLog, rlReport or other beaker libraries separately. 

So may I keep using the rhts- functions?

[the remark this answers is not in the document]
Comment (emanchado) [reply kselftests/include.sh]: Sure thing! If you think it's fine, it's probably fine. I just wanted to make sure we were not using deprecated APIs. Thanks for checking!

File: kselftests/include.sh (hunk 2 above)
Comment (emanchado) [reply kselftests/include.sh]: Oh, clever! I didn't know, that's pretty cool.

File: kselftests/include.sh (hunk 3 above)
Comment (emanchado) [reply kselftests/include.sh]: I mean that the variables at the top of `watch` have an indentation of 6 spaces, but the `eval` and the rest of the commands have an indentation of 8 spaces. Shouldn't they be aligned?

**Comment by emanchado** [comment]
BTW, I don't feel that great so I'll be taking today off. Maybe @rasibley can merge this if she approves?

File: kselftests/include.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (rasibley) [reply kselftests/include.sh]: @emanchado is right, we are working on removing legacy rhts functions and libraries from CKI tests to make the transition to testing in VMs smoother, in addition to using other test frameworks like TMT. See https://bugzilla.redhat.com/show_bug.cgi?id=1802610 for more of a background.

You can see all the changes needed, but basically we need to do the following:
- Convert all rhts-* commands to rstrnt-* compatible commands:  https://restraint.readthedocs.io/en/latest/commands.html
- Remove any rhts libraries, e.g.` [ -f /usr/bin/rhts_environment.sh ] `
- Remove rhts-make.include and rhts-lint mentioned in makefile, instead set the variables:
https://github.com/CKI-project/tests-beaker/pull/479/files#diff-3817e1d835f842835666e91e0fbbb14eR37

Fendy was also referencing cki_lib/libcki.sh library, but I don't think it's needed unless you are specifically taking advantage of the the following variables in your test: (@ftjahjad  is that right ? )
https://github.com/ftjahjad/tests-beaker/blob/f903aecceaed8b604195a1d5f77195f40fa6eaf3/cki_lib/libcki.sh#L11

Fendy's PR for reference: https://github.com/CKI-project/tests-beaker/pull/479/files

**Comment by rasibley** [comment]
@liuhangbin can't you combine this with the internal test wrapper ? Why do you need two separate tests ? I was hoping you would instead move your internal wrapper here and be able to test both RHEL and upstream kernels. Less work for you :-) Unless there is a specific reason for keeping them separate ?

File: kselftests/include.sh (hunk 3 above)
Comment (liuhangbin) [reply kselftests/include.sh]: Ah, I see.  I will align them

File: kselftests/include.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (ftjahjad) [reply kselftests/include.sh]: it is correct, most cki-beaker test use rhts-enviroment variables, 
https://github.com/ftjahjad/tests-beaker/blob/f903aecceaed8b604195a1d5f77195f40fa6eaf3/cki_lib/libcki.sh#L11 is setting beaker environment variables that deprecated  (rhts-environment.sh) in restrains.  
 
I think @rasibley  already covers everything, including  rhts-make.include and  rhts-<command> such as  replacing rhts-abort to   rstrnt-abort.

**Comment by liuhangbin** [comment]
> @liuhangbin can't you combine this with the internal test wrapper ? Why do you need two separate tests ? I was hoping you would instead move your internal wrapper here and be able to test both RHEL and upstream kernels. Less work for you :-) Unless there is a specific reason for keeping them separate ?

@rasibley The most codes are copied from the internal test. And I did some improvement.  The reason I didn't use internal wrapper is there are some internal links. like kselftest rpms links.

**Comment by rasibley** [comment]
> > @liuhangbin can't you combine this with the internal test wrapper ? Why do you need two separate tests ? I was hoping you would instead move your internal wrapper here and be able to test both RHEL and upstream kernels. Less work for you :-) Unless there is a specific reason for keeping them separate ?
> 
> @rasibley The most codes are copied from the internal test. And I did some improvement. The reason I didn't use internal wrapper is there are some internal links. like kselftest rpms links.

No problem, It should be possible to mask those internal links but we can do that later on, I don't want to block this. Thanks for all the improvements so far, I think it's ready to be merged, we can test it out further once we add it to kpet-db.

**Pull request merged**